tools/deploy/validate.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deploy
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/m3db/m3aggregator/aggregator"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3x/sync"
+)
+
+// Validator performs validation on a deployment target, returning nil if
+// validation succeeds, and an error otherwise.
+type Validator func() error
+
+// validatorFactory is a factory of validators.
+type validatorFactory interface {
+	// ValidatorFor generates a validator based on the instance,
+	// the instance group, and the instance target type.
+	ValidatorFor(
+		instance services.PlacementInstance,
+		group *instanceGroup,
+		targetType targetType,
+	) Validator
+}
+
+type factory struct {
+	client  AggregatorClient
+	workers xsync.WorkerPool
+}
+
+func newValidatorFactory(client AggregatorClient, workers xsync.WorkerPool) validatorFactory {
+	return factory{client: client, workers: workers}
+}
+
+func (f factory) ValidatorFor(
```

Comment:
Also, the interface itself is not exported so you can't actually use it outside the package. I tend to use public APIs in interfaces because if we ever want to export the interface in the future, you only need to change the interface name without touching the method names.